### ISSUE TITLE
Question: Storing all previous posts in of a feed in the config.

### ISSUE THREAD

**Issue by JoshuaCrewe** [FraGag/feeds-to-pocket#3]
I would be interested in hearing your reasoning for this decision, not that I think it is wrong, but I made my own decision when making this in Go.

When adding a feed `$ feeds-to-pocket ~/feeds-to-pocket.yaml add https://xkcd.com/atom.xml` this program can fetch all the previous blog posts and store them in the yaml file. When building this in Go, I found that when reading the URL for new posts I only needed to store the latest blog post each time. This means that the yaml file doesn't get bigger and bigger the longer this program is used.

I am not really familiar with Rust code (looking to learn) so I can't comment on the technical side of things but would it be possible to use a similar approach here?

To clarify : 

1. Request posts from RSS feed
2. Starting from the latest post from this request, check if it matches the url saved in the yaml file
3. If it does match, stop. We have the latest post (go to 5 else go to 4)
4. If it does not match move on the the next url in the feed. (Go to 3)
5. Update the yaml with the latest post from the feed

I don't know if that clarifies or not. Would love to hear your take on this decision and what led you to adopt the approach you did.

Thanks again!

**Comment by FraGag** [comment]
The reason I'm keeping the full history is that I've met some feeds in which entries regularly disappear before they reach the "bottom". One such feed is the [TASVideos Submissions](http://tasvideos.org/submissions.rss) feed, in which a submission will disappear when it's published.

Initially, I had planned on storing each feed's current entries (i.e. purging entries as they disappeared from the feed). That wouldn't have worked with the TASVideos Submissions feed, because an old entry can reappear at the bottom when another entry disappears. The application would then push that entry again to Pocket, and if you had archived the entry, that would "unarchive" it.

There's also the issue that I don't necessarily trust a feed to always add entries at the "top". This is more likely to happen to aggregator feeds, such as [begin end](https://www.beginend.net/), where some earlier entries can appear later if one of the source feeds is unavailable for a while, but the entries are still ordered chronologically, so with the strategy you describe, you could miss some entries.

As I was developing this, I was thinking that the application should probably have a mechanism for purging old entries, but I didn't implement this yet.

**Comment by JoshuaCrewe** [comment]
I see what you are saying. I thought there would be scenarios like this. I suppose the only real consideration is file size for the config? The only reason I was going in there was to remove feeds anyway which is no longer an issue with the work you just did over the weekend ?

The process of purging old entries would be to keep the "latest" 15 feeds or something with posts from more than six months ago being removed? Did you have a process in mind ?

**Comment by FraGag** [comment]
> I suppose the only real consideration is file size for the config?

Yes. In my case, my config is stored in my Dropbox, so every time it changes, it has to be reuploaded (though apparently, [Dropbox can upload deltas](https://www.dropbox.com/help/syncing-uploads/upload-entire-file) instead of the whole file each time, so it might not be so bad after all). For the record, my own config file currently stands at 639 KiB.

> Did you have a process in mind ?

I had a few ideas:

1. For each feed, keep _n_ times the number of entries the feed currently has, and purge the rest.
2. For each feed, keep _n_ entries more than the feed currently has, and purge the rest.
3. For each entry, keep track of when it was _last seen_, and purge the entries that were last seen earlier than some time relative to now (say, a month ago). This implies changing the file format a bit, but the application could migrate the config automatically. Checking when it was last seen means we won't purge an entry that's still in the feed (consider a feed with very infrequent activity). However, it also means that many timestamps all across the file will change every time the program is run (unless we cheat).

What do you think?

**Comment by JoshuaCrewe** [comment]
Gosh! When life gets busy busy!

I keep my feed in git (Sparkleshare) and is much smaller than yours, like 120kb or something. So it is not really that big a deal I suppose.

I don't think I am fully following the scenarios here.

1. (if _n_ is 30) Are we talking about a feed that has 100 posts keeping the latest 30 and discarding the older 70 ? Or something else?

2. This would have the config keep all the entries a feed has plus _n_ extra items that may have been removed from the site?

3. This sounds like the most complex but possibly closer to what I had in mind. Are we opening ourselves up for a world of pain by taking this route? How do you feel about this?

I need to check I am imagining the situation correctly. When this program runs it requests all the feeds from the RSS URL and checks each item against the yaml array in the config. If the item is not in the array then it pushes it to Pocket. So we _need_ to have a list of all items from the feed in the config to know which items are new? Is that about right?

Maybe this is just a non issue and we trade disk space for reliability? That is to say how much would we really be able to purge and still get a reliable result ?

**Comment by FraGag** [comment]
> 1. (if _n_ is 30) Are we talking about a feed that has 100 posts keeping the latest 30 and discarding the older 70 ? Or something else?

No, that was more like, say a feed has 30 entries, and I configure it to keep 2x the number of entries, then it would keep at most 60. This means that the number of extra entries that are kept in the file varies for each feed.

> 2. This would have the config keep all the entries a feed has plus _n_ extra items that may have been removed from the site?

Correct.

> 3. This sounds like the most complex but possibly closer to what I had in mind. Are we opening ourselves up for a world of pain by taking this route? How do you feel about this?

I don't know. This was actually the first approach I thought of, but then the more I thought about it, the less reliable it sounded. The scenario I imagined was a blog that had become inactive for a year, and then a post present in the feed was deleted. This would bring a _very old_ post back into the feed. Unless you configured a time delay so long it would basically be useless, this old post would be sent back to Pocket. This wouldn't happen with options 1 and 2.

> I need to check I am imagining the situation correctly. When this program runs it requests all the feeds from the RSS URL and checks each item against the yaml array in the config. If the item is not in the array then it pushes it to Pocket. So we _need_ to have a list of all items from the feed in the config to know which items are new? Is that about right?

That's about right. We need enough entries so that current entries are not sent twice and entries that might reappear later are not sent either.

(As an optimization, the application checks the YAML array _backwards_, i.e. from the most recent entry first, and exit as soon as a match is found. It's an optimization because current entries are more likely to be towards the end of the array. Still, the time it takes to check if an entry is new is proportional to the size of the history in the worst case.)

> Maybe this is just a non issue and we trade disk space for reliability? That is to say how much would we really be able to purge and still get a reliable result ?

This is in part why I haven't implemented a purge mechanism yet. I haven't yet felt the need to have the history be purged automatically, and I thought, if I want to purge it, I'll just do it manually (there's a good reason I chose a human-readable and human-editable file format).

Each feed has its own update frequency; some have regular updates, others have irregular updates. Some feeds have a lot of entries, others have very few. I don't think there is a perfect solution that isn't overly complex.

I think my current preference is towards option 2, keep _n_ extra items. But then, there are still choices to make:

* Should the number of extra entries to retain be global, be configurable per feed, or both?
* What's the default? Do we set a default value for _n_, or is the default not to purge?
* Is it OK to purge the "oldest" entries (as defined by their order of appearance in the YAML, considering that new entries are always added at the end), or do we need something more clever (actually keeping track of some date for each entry)?

**Comment by JoshuaCrewe** [comment]
I am with you. So it _sounds_ like this could be more trouble than it is worth. If we want to store a more or less complete list of a feeds posts  then we wont actually be shedding all that many entries. We might gain a bit in performance as there will be fewer entries to parse but this is going to minimal and bytes saved in space.

I am happy with not doing anything for this on the understanding that the gains are not enough for the solutions we have available. I mean to say that I feel like my original question has been answered and I am very grateful for the time you have taken to explain your thought process and implementation.

tldr;

My vote: do nothing. Your vote counts for more though and I am happy to help in any way that you need.

**Comment by FraGag** [comment]
I have no intention to work on this at this time. However, I will consider merging a pull request that implements a purge mechanism, so long as it considers the issues discussed above.
